src/web_app/js/peerconnectionclient.js

```diff
@@ -358,7 +361,100 @@ PeerConnectionClient.prototype.onRemoteStreamAdded_ = function(event) {
 };
 
 PeerConnectionClient.prototype.onError_ = function(tag, error) {
+  var localSdp = this.localSdp || null;
+  var remoteSdp = this.remoteSdp || null;
+  // Enumerate supported callstats error types.
+  // http://www.callstats.io/api/#enumeration-of-wrtcfuncnames
+  var webrtcFuncNames = {
+    getUserMedia: 'getUserMedia',
+    createOffer: 'createOffer',
+    createAnswer: 'createAnswer',
+    setLocalDescription: 'setLocalDescription',
+    setRemoteDescription: 'setRemoteDescription',
+    addIceCandidate: 'addIceCandidate'
+  };
+
   if (this.onerror) {
     this.onerror(tag + ': ' + error.toString());
+
+    // Only report supported error types to the callstats backend.
+    if (webrtcFuncNames[tag]) {
+      this.callStats_.reportError(this.pc_, this.conferenceId,
+        webrtcFuncNames[tag], error, localSdp, remoteSdp);
+    }
+  }
+};
+
+// Setup the callstats api and attach it to the peerconnection.
+PeerConnectionClient.prototype.setupCallStats_ = function() {
+  // Check dependencies.
+  if (typeof $ !== 'function' && typeof io !== 'function' &&
+      typeof jsSHA !== 'function')  {
+    trace('Callstats dependencies missing, stats will not be setup.');
+    return;
+  }
+
+  // Need to catch the error otherwise the peerConnection creation
+  // will fail.
+  try {
+    trace('Setting up callstats.');
+    // jscs:disable requireCapitalizedConstructors
+    /* jshint newcap: false */
+    var callStats = new callstats(null, io, jsSHA);
+    // jscs:enable requireCapitalizedConstructors
+    /* jshint newcap: true */
+
+    var appId = this.params_.callStatsParams.appId;
+    var appSecret = this.params_.callStatsParams.appSecret;
+    var userId = this.params_.clientId;
+    var conferenceId = this.params_.roomId;
+    // We do not know the remote client id. Setting NA for now.
+    var remoteUserId = 'NA';
+    // Multiplex should be used when sending audio and video on a peerConnection.
+    // http://www.callstats.io/api/#enumeration-of-fabricusage
+    // TODO: Might need to change this dynamically if an audio/video only call.
+    var usage = callStats.fabricUsage.multiplex;
+    var statsCallback = null;
+    var configParams = null;
+    var callback = function(status, msg) {
+      trace('Callstats status: ' + status + ' msg: ' + msg);
+    };
+
+    // Init the callstats api.
+    callStats.initialize(appId, appSecret, userId, callback, statsCallback,
+      configParams);
+
+    // Hookup the callstats api to the peerconnection object.
+    callStats.addNewFabric(this.pc_, remoteUserId, usage,
+      conferenceId, callback);
```

Comment:
there are perhaps two solutions:

1. setting up callstats just before createOffer may solve the issue of the
calls where the peer connection is created but callee did not join. This
will of course mess with our setup time, gathering time calculations for
the app. :(

2. An alternate solution, would be to setup callstats before onNN (like it
is now). However, explicitly call the fabricTerminated if the tab/call is
closed without the caller joining.

cc: @karthikbr82

On Tue, 16 Feb 2016 at 01:07, Justin Uberti <notifications@github.com>
wrote:

> In src/web_app/js/peerconnectionclient.js
> <https://github.com/webrtc/apprtc/pull/261#discussion_r52951984>:
>
> > +    // http://www.callstats.io/api/#enumeration-of-fabricusage
> > +    // TODO: Might need to change this dynamically if an audio/video only call.
> > +    var usage = callStats.fabricUsage.multiplex;
> > +    var statsCallback = null;
> > +    var configParams = null;
> > +    var callback = function(status, msg) {
> > +      trace('Callstats status: ' + status + ' msg: ' + msg);
> > +    };
> > +
> > +    // Init the callstats api.
> > +    callStats.initialize(appId, appSecret, userId, callback, statsCallback,
> > +      configParams);
> > +
> > +    // Hookup the callstats api to the peerconnection object.
> > +    callStats.addNewFabric(this.pc_, remoteUserId, usage,
> > +      conferenceId, callback);
>
> I wonder if we could simply generate a pseudo-id based on the conference
> name. E.g. caller id becomes "-0", callee id is "-1". Then that would be
> the same on both sides and could be used for correlation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/webrtc/apprtc/pull/261/files#r52951984>.
>